### PR TITLE
[ci skip] in all pr commit message and title

### PR DESCRIPTION
**Pull request by francesco086**

### CODE REVIEW

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
🙏 
IMHO I think this should work.
I can not see the scenario a user would like to avoid `[skip ci]`

However according to [this](https://github.com/iterative/cml/issues/1023#issuecomment-1140154009) the addition of `[skip ci]` should be determined with an input parameter within the command

File: src/cml.js [hunk 1]
```diff
@@ -463,15 +463,12 @@ class CML {
...
-      if (!(merge || rebase || squash)) {
-        commitMessage += ' [skip ci]';
-      }
+      let commitMessage = `CML PR for ${shaShort} [ci skip]`;
       await exec(`git commit -m "${commitMessage}"`);
       await exec(`git push --set-upstream ${remote} ${source}`);
     }
 
-    const title = `CML PR for ${target} ${shaShort}`;
+    const title = `CML PR for ${target} ${shaShort} [ci skip]`;
```
[the remark this answers is not in the document]
Comment (francesco086) [reply src/cml.js:471]: https://github.com/iterative/cml/issues/1023#issuecomment-1141205374

[the remark this answers is not in the document]
Comment (francesco086) [reply src/cml.js:471]: I wanted to show you an example, and by doing that I realized that I overlooked it.
https://gitlab.com/francesco-calcavecchia/titanic-demo-dvc-as-model-registry/-/commit/936744e6af86c30c394e126d168240989d3133c0

The MR title is not included in the commit message, but only in the commit note. Damn... ok, then I remove this change.

[the remark this answers is not in the document]
Comment (francesco086) [reply src/cml.js]: Removed :)

**Comment by francesco086** [comment]
I tried to follow your instructions, see if I missed something

File: src/cml.js [hunk 2]
```diff
@@ -464,7 +465,7 @@ class CML {
       await exec(`git checkout -b ${source}`);
       await exec(`git add ${paths.join(' ')}`);
       let commitMessage = `CML PR for ${shaShort}`;
-      if (!(merge || rebase || squash)) {
+      if ( skipCI ){
```
Comment (DavidGOrtega) [inline src/cml.js:468]: We want cml pr to still do skipCI if not automerge

```suggestion
      if ( skipCI ||  !(merge || rebase || squash)){
```

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]

**Comment by DavidGOrtega** [comment]
francesco because we have changed a binary we might redo a template test.
Just replace in ./bin/cml/pr.js 

```js
expect(output).toMatchInlineSnapshot(`
      "cml.js pr <glob path...>

      Create a pull request with the specified files

      Options:
        --help                 Show help                                     [boolean]
        --version              Show version number                           [boolean]
        --log                  Maximum log level
                [string] [choices: \\"error\\", \\"warn\\", \\"info\\", \\"debug\\"] [default: \\"info\\"]
        --md                   Output in markdown format [](url).            [boolean]
        --merge, --auto-merge  Try to merge the pull request upon creation.  [boolean]
        --rebase               Try to rebase-merge the pull request upon creation.
                                                                             [boolean]
        --squash               Try to squash-merge the pull request upon creation.
                                                                             [boolean]
        --remote               Sets git remote.           [string] [default: \\"origin\\"]
        --user-email           Sets git user email.
                                             [string] [default: \\"olivaw@iterative.ai\\"]
        --user-name            Sets git user name.   [string] [default: \\"Olivaw[bot]\\"]
        --repo                 Specifies the repo to be used. If not specified is
                               extracted from the CI ENV.                     [string]
        --token                Personal access token to be used. If not specified in
                               extracted from ENV REPO_TOKEN.                 [string]
        --driver               If not specify it infers it from the ENV.
                                   [string] [choices: \\"github\\", \\"gitlab\\", \\"bitbucket\\"]"
    `);
```
with
```js
expect(output).toMatchInlineSnapshot()
```

and then run

```sh
npm run do_snapshots
```

Last thing! Promise! 🙏

**Comment by francesco086** [comment]
That does a lot of sense, but then I need instructions on how to set up npm for this project, keep in mind that I have never worked with js before...

Btw I must say that I could not understand why `!(merge | rebase | squash)`. What is the sense? 
If you don't want to automerge then you do want to run CI? I must say I don't understand it :(

Here we are controlling if CI runs on the new branch for the new commit created for, say, the new dvc.lock. It has nothing to do with master. Correct?

File: bin/cml/pr.js [hunk 3]
```diff
@@ -19,6 +19,10 @@ exports.builder = (yargs) =>
         type: 'boolean',
         description: 'Output in markdown format [](url).'
       },
+      skipCI: {
+        type: 'booelan',
```
Comment (casperdcl) [inline bin/cml/pr.js:23]:
```suggestion
        type: 'boolean',
```

File: src/cml.js [hunk 4]
```diff
@@ -464,7 +465,7 @@ class CML {
       await exec(`git checkout -b ${source}`);
       await exec(`git add ${paths.join(' ')}`);
       let commitMessage = `CML PR for ${shaShort}`;
-      if (!(merge || rebase || squash)) {
+      if ( skipCI ||  !(merge || rebase || squash)){
```
Comment (casperdcl) [inline src/cml.js:468]:
```suggestion
      if (skipCI || !(merge || rebase || squash)){
```


but as per the last line in [this comment](https://github.com/iterative/cml/issues/1023#issuecomment-1140154009) I wonder if `skipCI` should mean ` [skip ci]` is added to `title` rather than `commitMessage`.

File: bin/cml/pr.js [hunk 5]
```diff
@@ -19,6 +19,10 @@ exports.builder = (yargs) =>
...
+        description: 'Skip CI for the commit created for the pull request?'
```
Comment (casperdcl) [inline bin/cml/pr.js:24]:
```suggestion
        description: 'Force skip CI for the created commit (if any)'
```

**Review by casperdcl — CHANGES_REQUESTED** [review]

**Comment by DavidGOrtega** [comment]
> Btw I must say that I could not understand why `!(merge | rebase | squash)`. What is the sense? If you don't want to automerge then you do want to run CI? I must say I don't understand it :(

Not everyone might want to do auto merge. Indeed I prefer to review the PR. Also if in one experiment branch we commit X times parameters we will generate X PRs and we can review and pick the very best one. Automerge is not desired in every scenario

**Comment by francesco086** [comment]
> 

I don't argue on that, my question is: why would you skip CI just because you did not set to auto-merge?

**Comment by DavidGOrtega** [comment]
> I don't argue on that, my question is: why would you skip CI just because you did not set to auto-merge?

You might have events based on push running the CI again and again...

File: src/cml.js [hunk 6]
```diff
@@ -464,7 +465,7 @@ class CML {
       await exec(`git checkout -b ${source}`);
       await exec(`git add ${paths.join(' ')}`);
       let commitMessage = `CML PR for ${shaShort}`;
-      if (!(merge || rebase || squash)) {
+      if (skipCI || !(merge || rebase || squash)) {
```
Comment (casperdcl) [inline src/cml.js:468]: per the last line in [this comment](https://github.com/iterative/cml/issues/1023#issuecomment-1140154009) I wonder if `skipCI` should mean ` [skip ci]` is added to `title` rather than `commitMessage`?

Comment (francesco086) [reply src/cml.js:468]: I suppose you refer to `option to add [skip ci] to the PR title (afaik the PR title usually becomes the commit title upon merge)`?

I checked and on GitLab it won't work. The commit message is something like:
"Merge branch 'feature-branch-cml-pr-10d62239' into 'feature-branch''"

The proper way would be to go the way I sketched [here](https://github.com/iterative/cml/issues/1023#issuecomment-1141277204).

**Comment by francesco086** [comment]
> > I don't argue on that, my question is: why would you skip CI just because you did not set to auto-merge?
> 
> You might have events based on push running the CI again and again...

and you think that if one set auto-merge he/she wants that?

Currently (before this PR):
- if auto-merge is not set then skip CI
- if auto-merge is set then don't skip CI

I don't follow the logic.

File: src/cml.js (hunk 6 above)
Comment (francesco086) [reply src/cml.js:468]: And btw, as of now, even with the current PR, the issue would still be open. The only way to avoid undesired CI runs on the feature-branch after the merge is to avoid setting auto-merge, and manually add [skip ci] to the commit message before merging the PR created by CML.

Comment (DavidGOrtega) [reply src/cml.js:468]: Right now it looks good.
GH and GL will honor skip ci in the commit message.

**Comment by DavidGOrtega** [comment]
> Currently (before this PR):
if auto-merge is not set then skip CI
if auto-merge is set then don't skip CI
I don't follow the logic.


OK. I will try to explain 🙂 

The PR to me looks good. 
cml pr was conceptually created to create a PR against an experiment branch. The problem is that many workflows reacts on simple push event and/or create_pull_request, so when `cml pr` acts it creates a new branch and pushes it creating a new PR, that will generate another triggers (two in the case that push and create_pull_request are enabled) if `skip ci` is not set we are going to enter in a loop of PR generation.

On the other hand auto-merge might need a run of the workflow because you might want to do a check with the merged commit. Or maybe not because you might end up within a loop.

Thats why we actually need to setup the `skip ci` understanding that when creating the branch and creating the PR we must always skip to avoid entering in a recursive loop. 
The parameter that you created is to control the auto-merge possibility of skipping the ci

Does this makes sense?

**Review by DavidGOrtega — APPROVED** [review]

**Comment by casperdcl** [comment]
> `cml pr --merge` then:
> 
> 1. CI runs on the new branch created (let's call it `cml-branch-1`), generating in turn a new branch `cml-branch-2`, etc.
> 2. once the branch `cml-branch-1` is automatically merged into the original branch (typically a feature branch), CI will run again, which will again create a new branch `cml-branch-1b`. This branch will in turn generate a new look as in 1.

I don't follow (branch names are confusing). Do you mean "2. once `cml-branch-2` is automatically merged into `cml-branch-1` ..." instead? If so, the flow would be:

```mermaid
flowchart LR
M(main) -->|1. branch| F(feat) -->|2. branch| C1(cml-branch-1) -->|3. CI: cml pr --merge| C2(cml-branch-2)
C2 -->|4. CI: auto merge| C1 -->|5. CI: cml pr --merge| C1b(cml-branch-1b) -->|CI: infinite loop| C1
```

If so, this makes sense. And you ideally would like `cml-branch-1b` never to be created, right?

> `cml pr` (without the auto-merge), then:
> 
> 1. CI will not run on the branch `cml-branch-1`. So far so good!

I don't follow.

> 2. Once I will manually merge `cml-branch-1` back in the feature branch, if I don't manually add "[ci skip]" to the commit message, CI will run again on the feature branch, generating `cml-branch-1b`.


```mermaid
flowchart LR
M(main) -->|1. branch| F(feat) -->|2. branch| C1(cml-branch-1) -->|3. CI: cml pr| C2(cml-branch-2)
C2 -->|"4. manual merge, entering [skip ci] into message"| C1
```

Anyway there are many work-arounds:

1. > deterministic runs [not opening PRs because of] diff being empty.
2. > feature request: make "[skip ci]" [appear in] all commit message of branches/PR created via `cml pr`
3. feature request:

```mermaid
flowchart LR
M(main) -->|1. branch| F(feat) -->|2. branch| C1(cml-branch-1) -->|3. CI: cml pr --merge --skip-ci| C2("cml-branch-2 with commit containing [skip ci]")
C2 -->|"4. auto merge without any CI tests running"| C1
```

4. [feature request](https://github.com/iterative/cml/issues/1023#issuecomment-1141277204):

```mermaid
flowchart LR
M(main) -->|1. branch| F(feat) -->|2. branch| C1(cml-branch-1) -->|3. CI: cml pr --merge --skip-ci| C2(cml-branch-2)
C2 -->|"4. auto merge, inserting [skip ci] into the merge commit message"| C1
```

So basically the problem with this PR currently is "auto merge without any CI tests running" right? Work-around (3) is broken? So we should do work-around (4)?

**Comment by casperdcl** [comment]
Looks great, @francesco086 - think CLI UX can stay the same, it's just the `[skip ci]` needs to be in the MR title (at least on GitLab) rather than commit message. PRs welcome!
